### PR TITLE
Bugfix: util: Correctly handle Number value types in GetArg/GetBoolArg

### PR DESCRIPTION
**Pull request by luke-jr**
Currently, Number values can cause ~~GetArg/~~ GetBoolArg to throw an exception, which most of the code isn't expecting (eg #24457).

AFAIK it isn't possible to normally get Numbers in settings right now, but ~~that's expected to change with #15936 (which could create downgrading issues without this fixed), and in any case~~ it isn't a great idea to randomly crash because of a logically-value settings.json value anyway.

This *doesn't* fix Get\*Arg from throwing in the case it encounters an Array or Object. It's not obvious how to handle those scenarios. However, the function~~s~~ touched are refactored to explicitly clarify that those scenarios will throw.

~~NOTE: Based on 0.20 branch-point, so a clean merge to everything newer. For testing, you will probably want to merge onto 23.x or master.~~

### CODE REVIEW

**Review by ghost — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/24479/commits/3766c66b4f6126c30a254a9c0f61602d7a42eaaf

Agree with the concept and couldn't find any issues with the code.

**Comment by luke-jr** [comment]
Rebased due to merge of partial fix in #24498

**Comment by fanquake** [comment]
The unit tests are failing:
```bash
test/getarg_tests.cpp(103): error: in "getarg_tests/setting_args": exception std::runtime_error expected but not raised
test/getarg_tests.cpp(104): error: in "getarg_tests/setting_args": exception std::runtime_error expected but not raised
test/getarg_tests.cpp(110): error: in "getarg_tests/setting_args": exception std::runtime_error expected but not raised
test/getarg_tests.cpp(111): error: in "getarg_tests/setting_args": exception std::runtime_error expected but not raised
test/getarg_tests.cpp(117): error: in "getarg_tests/setting_args": exception std::runtime_error expected but not raised
test/getarg_tests.cpp(118): error: in "getarg_tests/setting_args": exception std::runtime_error expected but not raised
test/getarg_tests.cpp(53): Leaving test case "setting_args"; testing time: 63184us
```

**Comment by luke-jr** [comment]
Fixed tests.

File: src/util/system.cpp [hunk 1]
```diff
@@ -603,7 +603,21 @@ int64_t ArgsManager::GetIntArg(const std::string& strArg, int64_t nDefault) cons
...
+        case UniValue::VOBJ:
+        case UniValue::VARR:
+            // Throws an exception
+            value.get_str();
+            assert(false);
+        case UniValue::VSTR:
+        case UniValue::VNUM:
+            return InterpretBool(value.getValStr());
+    }
+    assert(false);
```
Comment (ryanofsky) [inline src/util/system.cpp:620]: In commit "Bugfix: util: Correctly handle Number value types in GetArg/GetBoolArg" (33722279495ebf566436a274765068217b2d3543)

This is pretty verbose and a little hard to follow. Could consider stealing Marco's style from https://github.com/bitcoin/bitcoin/pull/24498#discussion_r822889392 and writing something like:

```c++
return value.isNull() ? fDefault :
       value.isBool() ? value.get_bool() :
       value.isNum()  ? InterpretBool(value.getValStr()) :
                        InterpretBool(value.get_str());
```

**Review by ryanofsky — APPROVED** [review]
I could see people having different opinions about this change. Personally I wouldn't object to it, but I also don't think it is an improvement overall.

The PR has changed since it was opened, but now it is pretty simple. Right now, if a user edits their `settings.json` file, or if the `settings.json` file gets corrupted some other way so it contains a numeric int or float value where a boolean value is expected, the current `GetBoolArg` method will throw an exception. This PR changes `GetBoolArg` behavior in that case to return `false` if the number is zero and `true` if the number is nonzero instead of throwing an exception.

Since this PR does not change behavior in other cases, `GetBoolArg` still throws exceptions if the settings value is an array or object, and `GetBoolArg` still has strange behavior for string values of returning true for the empty string `""`, true for strings which are non-zero decimal numbers, and false for all other strings.

Since this PR isn't removing every case where `GetBoolArg` throws exceptions, and since specifying a number where a boolean value is expected is a *real* indication of corruption or misconfiguration, I think throwing an exception is probably the best thing method can do right now to deal with whatever bad value is causing the problem. In longer run, if a change like #16545 goes through, settings could be typed, and errors could be triggered earlier on startup before an exception could happen here.

Anyway, this is not a very strong opinion, and I'm fine with other approaches.

Code review ACK 33722279495ebf566436a274765068217b2d3543

File: src/util/system.cpp (hunk 1 above)
Comment (luke-jr) [reply src/util/system.cpp:620]: Huh? I changed it to a switch because it's very readable and the repeated trinary operator is hard to follow...

**Comment by achow101** [comment]
The PR didn't seem to attract much attention in the past. Also, the issue seems not important enough right now to keep it sitting around idle in the list of open PRs.

Closing due to lack of interest.
